src/main/java/build/buildfarm/instance/memory/MemoryInstance.java

```diff
@@ -253,30 +274,32 @@ protected void enqueueOperation(Operation operation) {
 
   @Override
   protected void updateOperationWatchers(Operation operation) throws InterruptedException {
-    synchronized(watchers) {
-      List<Predicate<Operation>> operationWatchers =
-          watchers.get(operation.getName());
-      if (operationWatchers != null) {
-        if (operation.getDone()) {
-          watchers.remove(operation.getName());
-        }
-        long unfilteredWatcherCount = operationWatchers.size();
-        super.updateOperationWatchers(operation);
-        ImmutableList.Builder<Predicate<Operation>> filteredWatchers = new ImmutableList.Builder<>();
-        long filteredWatcherCount = 0;
-        for (Predicate<Operation> watcher : operationWatchers) {
-          if (watcher.test(operation)) {
-            filteredWatchers.add(watcher);
-            filteredWatcherCount++;
+    super.updateOperationWatchers(operation);
+
+    Set<Predicate<Operation>> operationWatchers = watchers.get(operation.getName());
+    synchronized (watchers) {
+      for (Predicate<Operation> watcher : operationWatchers) {
+        ListenableFuture<Boolean> stillWatchingFuture = watcherService.submit(new Callable<Boolean>() {
+          @Override
+          public Boolean call() {
+            return watcher.test(operation) && !operation.getDone();
```

Comment:
Do I understand correctly that this will perform the check+remove only once, but asynchronously? What's the point to doing it asynchronously, then -- if this is called before the operation is completed, then there's an inherent race, and if it's after, then wouldn't doing it synchronously be the same?

Reply:
The point to the asynchronous here is to limit time under `synchronized (watchers)` - the watcher.test is potentially a lengthy operation - and to disconnect the caller's responsibility to scale with the universal watcher set. Since this synchronization occurs without respect to the operation, queuing the updates for the watchers into an executor service is necessary to avoid undue sequencing of updates. I've seen needless pid and memory exhaustion while waiting for synchronous access to deliver updates during peak execution traffic.

There is no race as far as the watcher is concerned here - this is an update observer, whose delivery content is immutable, and whenever it manages to get the operation update with `done` of `true` out, the watcher proceeds, no matter how much the actual delivery lags. At worst, we'll miss the opportunity to skip an update because we've already got a further state queued.